### ISSUE TITLE
[Feature] Integrity Checks / Cryptographic Verification

### ISSUE THREAD

**Issue by hendursaga** [violentmonkey/violentmonkey#1558]
**Motivation**

It appears @require does not require or allow any sort of cryptographic verification of any external scripts or even a simple integrity check, akin to Subresource Integrity.

**Proposed Solution**

I do not know if there is any formal specification for userscripts, or if it's just convention or whatnot, but at least allowing script authors to include a SHA256 or better hash for each @require so that users are less vulnerable would be a minimal solution. Past that, I'm not sure what the best route to take for cryptographic verification, such as with PGP, OpenBSD's signify, SSH keys, or whatnot, would be.

**Comment by tophf** [comment]
We can do it just like Tampermonkey: https://www.tampermonkey.net/documentation.php?ext=dhdg#_require

**Comment by rezad1393** [comment]
maybe someone can host a ipfs service and pin the needed libraries.
that is the best long term for at least two issues.

first is the availability
for example right now this library hosted by this website is down:
https://cdn.bootcdn.net/ajax/libs/jquery/3.4.1/jquery.slim.min.js

second
integrity via ipfs.

**Comment by gera2ld** [comment]
I don't think IPFS has anything to do with this.
- Public gateway is slow and may time out if the file is not accessed by anyone for a while. So it's less stable.
- If all you care about is the hash, you can `@require` an IPFS URL directly.

**Comment by rezad1393** [comment]
> I don't think IPFS has anything to do with this.
> 
>     * Public gateway is slow and may time out if the file is not accessed by anyone for a while. So it's less stable.
> 
>     * If all you care about is the hash, you can `@require` an IPFS URL directly.

"best long term"

I didnt mean to move all the things to ipfs.
I just wanted to show one protocol that is doing these cool things that are needed for integrity.

when you get a file from url you still need someone to host it. that property is shared between ipfs and url/http.
so that someone can use ipfs and pin that file. 

I have heard that "if the file is not accessed by anyone for a while" but that argument thinks no one will host (pin) that file long term.
in that worst case scenario then user can use another host or even selfhost the file by finding it in other places via its hash.
I cant say the same thing about the library that right now is down for me.

the hash + url is a good integrity solution but not that much can be done when file is gone (host is down).
but in ipfs if I had downloaded that lib before I can just use that cache version.
because I would not have to check a website each time an script check for updates and uses a @require lib that is down.

long story short, even though ipfs is advertises as kind of git+BitTorrent (which it is) I really like its p2p and cache ability.

imagine for a moment all the libraries font css and so on  hosted by google and other cooperation that you use daily.
browser cache in not really good for caching these cross-domain files , but ipfs is.
say tomorrow google goes crazy and host an ipfs instance for these files. I would be able to use those file way more securely and don't have to trust google at all. and I can use ipfs cache and if browser support for ipfs (which is considered in firefox and done in brave) then ipfs.google.com/hashfile can be hosted by anyone with hashfile and google can be last resort but with integrity built-in.

I think this is too off-topic and I am sorry if I rambled.

**Comment by gera2ld** [comment]
I love IPFS too but I think it's for a different scenario. I'll hide the comments since they are off-topic, but feel free to continue in Discussions.

**Comment by rezad1393** [comment]
> I love IPFS too but I think it's for a different scenario. I'll hide the comments since they are off-topic, but feel free to continue in Discussions.

oh.
I was trying to convert a believer :)

**Comment by andrewsuzuki** [comment]
Copying over notes from the duplicate issue I created:

Tampermonkey uses a URL fragment (e.g. `jquery-2.1.1.min.js#sha-512=23fd34...`). I don't particularly like this solution because some resources might need the fragment themselves.

Something like this might be cleaner: `jquery-2.1.1.min.js integrity:sha512-23fd34...` (url, space, `integrity:` followed by [the same format](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) that `<script />` uses).

Can use SubtleCrypto [digest()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest) to generate hashes (supports `SHA-1`, `SHA-256`, `SHA-384`, `SHA-512`). However, I don't think that would work in insecure contexts.

**Comment by tophf** [comment]
Maybe omit `integrity` and just used the hash name: `jquery-2.1.1.min.js sha512:23fd34...`?

> SubtleCrypto 
> insecure contexts

It's ok because the extension context is secure, which is where the resource is downloaded and checked.

**Comment by andrewsuzuki** [comment]
That would work. My line of reasoning was: `<script />` uses a dash rather than a colon, so maintain the same format. Then `integrity:` would allow other named parameters/flags for require/resource in the future, though to be fair I can't think of anything else that would go there.

**Comment by mitchellmebane** [comment]
@andrewsuzuki do you have any particular scenarios in mind where "some resources might need the fragment themselves"? AFAIK fragments aren't sent to the server, so I can't think of any case where that would be possible.

**Comment by andrewsuzuki** [comment]
@mitchellmebane Ah, you're right

**Comment by Vinfall** [comment]
Tampermonkey also supports `@resource` according to their [Subresource Integrity](https://www.tampermonkey.net/documentation.php#api:Subresource_Integrity) documentation, although this one is less useful than `@require`.

While I think it's fine to use their fragment format, if you have to make this incompatible, maybe you can try use their `SRIsecured` flag. Something like this (ignore unchanged lines):

```diff
// @resource SRIsecured1 http://example.com/favicon1.ico#md5=ad34bb...
// @resource SRIsecured2 http://example.com/favicon2.ico#md5=ac3434...,sha256=23fd34...
-// @require              https://code.jquery.com/jquery-2.1.1.min.js#md5=45eef...
-// @require              https://code.jquery.com/jquery-2.1.2.min.js#md5-ac56d...,sha256-6e789...
+// @require SRIsecured https://code.jquery.com/jquery-2.1.1.min.js#md5=45eef...
+// @require SRIsecured https://code.jquery.com/jquery-2.1.2.min.js#md5-ac56d...,sha256-6e789...
// @require              https://code.jquery.com/jquery-3.6.0.min.js#sha256-/xUj+3OJU...ogEvDej/m4=
```

**Comment by sken130** [comment]
Could this feature be put into higher priority? Cybersecurity is very important these days, and we need some measure to migitate CDN hijacks.
Please support common hashes such as SHA-256, SHA-384 and SHA-512, thanks.

**Comment by Rplus** [comment]
💪

**Comment by kedryte** [comment]
> Tampermonkey uses a URL fragment (e.g. `jquery-2.1.1.min.js#sha-512=23fd34...`). I don't particularly like this solution because some resources might need the fragment themselves.
> 
> Something like this might be cleaner: `jquery-2.1.1.min.js integrity:sha512-23fd34...` (url, space, `integrity:` followed by [the same format](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) that `<script />` uses).

I don't agree with this for a couple reasons.

1) I don't think it's true that some resources might need the fragment, because the fragment is by design not even passed to the server when requesting a resource, it's only available inside the browser via JS.

2) I think it's important to maintain cross-compatibility with different userscript extensions that may not support this feature. Using the # fragment is a good way to add extra security on extensions that support it without breaking the ones that don't.

I think Tampermonkey does it the right way overall. Using the same format as `<script/>` is important to minimize confusion.

**Comment by deevroman** [comment]
A nice side effect is that when reinstalling the script, you don't have to redownload dependencies if their hash hasn't changed. Currently, because downloads are sequential, even with a large list of small dependencies, downloading them can take more than 10 seconds, which slightly degrades the user/dev experience.

**Comment by rezad1393** [comment]
> A nice side effect is that when reinstalling the script, you don't have to redownload dependencies if their hash hasn't changed. Currently, because downloads are sequential, even with a large list of small dependencies, downloading them can take more than 10 seconds, which slightly degrades the user/dev experience.

specially when most scripts have changes of script inner code itself and not the dependencies.
